runtime/cudaq/platform/mqpu/MultiQPUPlatform.cpp

```diff
@@ -9,100 +9,126 @@
-  void resetExecutionContext() override {
-    cudaq::info("MultiQPUPlatform::resetExecutionContext QPU {}", qpu_id);
-    auto tid = std::hash<std::thread::id>{}(std::this_thread::get_id());
-    auto ctx = contexts[tid];
-    handleObservation(ctx);
-    cudaq::getExecutionManager()->resetExecutionContext();
-    contexts[tid] = nullptr;
-    contexts.erase(tid);
-  }
-};
+  bool supports_task_distribution() const override { return true; }
 
-class MultiQPUQuantumPlatform : public cudaq::quantum_platform {
-public:
-  ~MultiQPUQuantumPlatform() = default;
-  MultiQPUQuantumPlatform() {
-    int nDevices;
-    cudaGetDeviceCount(&nDevices);
-
-    auto envVal = spdlog::details::os::getenv("CUDAQ_MQPU_NGPUS");
-    if (!envVal.empty()) {
-      int specifiedNDevices = 0;
-      try {
-        specifiedNDevices = std::stoi(envVal);
-      } catch (...) {
+  void setTargetBackend(const std::string &description) override {
+    const auto getOpt = [](const std::string &str,
+                           const std::string &prefix) -> std::string {
+      const auto prefixPos = str.find(prefix);
+      if (prefixPos == std::string::npos)
+        return "";
+      const auto endPos = str.find_first_of(";", prefixPos);
+      return (endPos == std::string::npos)
+                 ? str.substr(prefixPos + prefix.size() + 1)
+                 : cudaq::split(str.substr(prefixPos + prefix.size() + 1),
+                                ';')[0];
+    };
+
+    if (description.find("remote-execution") != std::string::npos) {
+      if (!cudaq::registry::isRegistered<cudaq::QPU>("RemoteSimulatorQPU"))
         throw std::runtime_error(
-            "Invalid CUDAQ_MQPU_NGPUS environment variable, must be integer.");
+            "Unable to retrieve RemoteSimulatorQPU implementation.");
+      auto urls = cudaq::split(getOpt(description, "url"), ',');
+      auto sims = cudaq::split(getOpt(description, "backend"), ',');
+      const bool autoLaunch =
+          description.find("auto_launch") != std::string::npos;
+
+      const auto formatUrl = [](const std::string &url) -> std::string {
+        auto formatted = url;
+        if (formatted.rfind("http", 0) != 0)
+          formatted = std::string("http://") + formatted;
+        if (formatted.back() != '/')
+          formatted += '/';
+        return formatted;
+      };
+
+      if (autoLaunch) {
+        urls.clear();
+        if (sims.empty())
+          sims.emplace_back("qpp");
+        const int numInstances = std::stoi(getOpt(description, "auto_launch"));
+        cudaq::info("Auto launch {} REST servers", numInstances);
+        for (int i = 0; i < numInstances; ++i) {
+          m_remoteServers.emplace_back(
+              new cudaq::AutoLaunchRestServerProcess());
+          urls.emplace_back(m_remoteServers.back()->getUrl());
+        }
+        // Allows some time for the servers to start
+        std::this_thread::sleep_for(std::chrono::milliseconds(1000));
```

Comment:
I'm curious how this number (1 second) was chosen. Do we expect this number to need to be tuned at some point in the future depending on where the remote RESET server is running? (I.e., what happens if you have to wait for a job in a busy job queue to start?)

Reply:
I'll add that in the old `qpud-client` I had set something similar here and @boschmitt reported issues with the client not being able to find the server in time. So this timing may come back to haunt us. I wonder if there is a way to have the client wait for a server message instead?

Reply:
I've updated this logic in https://github.com/NVIDIA/cuda-quantum/pull/1012/commits/7ea2535c183b04f48e3c10538ebc0e25b0c8c330 to ping the server (with a timeout limit).